mmcv/runner/hooks/optimizer.py

```diff
@@ -34,128 +42,237 @@ def after_train_iter(self, runner):
-
-    Refer to https://arxiv.org/abs/1710.03740 for more details.
-
-    Args:
-        loss_scale (float | str | dict): Scale factor multiplied with loss.
-            If loss_scale is a float, static loss scaling will be used with
-            the specified scale. If loss_scale is a string, it must be
-            'dynamic', then dynamic loss scaling will be used.
-            It can also be a dict containing arguments of LossScaler.
-            Defaults to 512.
-    """
-
-    def __init__(self,
-                 grad_clip=None,
-                 coalesce=True,
-                 bucket_size_mb=-1,
-                 loss_scale=512.,
-                 distributed=True):
-        self.grad_clip = grad_clip
-        self.coalesce = coalesce
-        self.bucket_size_mb = bucket_size_mb
-        self.distributed = distributed
-        if loss_scale == 'dynamic':
-            self.loss_scaler = LossScaler(mode='dynamic')
-        elif isinstance(loss_scale, float):
-            self.loss_scaler = LossScaler(init_scale=loss_scale, mode='static')
-        elif isinstance(loss_scale, dict):
-            self.loss_scaler = LossScaler(**loss_scale)
-        else:
-            raise ValueError('loss_scale must be of type float, dict, or '
-                             f'"dynamic", got {loss_scale}')
-
-    def before_run(self, runner):
-        """Preparing steps before Mixed Precision Training.
-
-        1. Make a master copy of fp32 weights for optimization.
-        2. Convert the main model from fp32 to fp16.
-        """
-        # keep a copy of fp32 weights
-        old_groups = runner.optimizer.param_groups
-        runner.optimizer.param_groups = copy.deepcopy(
-            runner.optimizer.param_groups)
-        state = defaultdict(dict)
-        p_map = {
-            old_p: p
-            for old_p, p in zip(
-                chain(*(g['params'] for g in old_groups)),
-                chain(*(g['params'] for g in runner.optimizer.param_groups)))
-        }
-        for k, v in runner.optimizer.state.items():
-            state[p_map[k]] = v
-        runner.optimizer.state = state
-        # convert model to fp16
-        wrap_fp16_model(runner.model)
-
-    def copy_grads_to_fp32(self, fp16_net, fp32_weights):
-        """Copy gradients from fp16 model to fp32 weight copy."""
-        for fp32_param, fp16_param in zip(fp32_weights, fp16_net.parameters()):
-            if fp16_param.grad is not None:
-                if fp32_param.grad is None:
-                    fp32_param.grad = fp32_param.data.new(fp32_param.size())
-                fp32_param.grad.copy_(fp16_param.grad)
-
-    def copy_params_to_fp16(self, fp16_net, fp32_weights):
-        """Copy updated params from fp32 weight copy to fp16 model."""
-        for fp16_param, fp32_param in zip(fp16_net.parameters(), fp32_weights):
-            fp16_param.data.copy_(fp32_param.data)
+if TORCH_VERSION != 'parrots' and TORCH_VERSION >= '1.6.0':
```

Comment:
Last comment: The indentation is a little bit annoying. Can we create two hooks first, namely `PT16Fp16OptimizerHook` and `PT15Fp16OptimizerHook`, then assign one of them to `Fp16OptimizerHook` according to the version?